Modify ensip 16 (#41)
ensdomains/ensips: ensips/x.md

## Specification

### Events

#### Registry Events

```solidity
// Emitted when a new subname is registered
event NewSubname(bytes32 indexed labelhash, string label);

// Standard ERC1155 transfer event for name ownership changes
event TransferSingle(
    address indexed operator,
    address indexed from,
    address indexed to,
    uint256 id,
    uint256 value
);

// Standard ERC1155 transfer event for multiple name ownership changes
event TransferBatch(
  address indexed operator,
  address indexed from,
  address indexed to,
  uint256[] ids,
  uint256[] values
)

// Standard ERC721 transfer event for name ownership changes
event Transfer(
  address indexed from,
  address indexed to,
  uint256 tokenId,
)

// Emitted when subregistry is updated
event SubregistryUpdate(
    address indexed registry,
    uint256 indexed id,
    address subregistry
);

// Emitted when resolver is updated
event ResolverUpdate(
    address indexed registry,
    uint256 indexed id,
    address resolver

> );

Reviewer: In that case I think it would be good to use the existing `NewResolver` event in the Registry: 
```solidity    
// Logged when the resolver for a node changes.
event NewResolver(bytes32 indexed node, address resolver);
```

> );

Author: I am not sure what you are suggesting here.

```

NOTE: The registry contract supports either ERC721 (Transfer) or ERC1155 (TransferSingle, TransferBatch)

#### Resolver Events

To keep compatibility with ENS v1, resolver events maintain the same event interface as ENS v1 with one notable difference.

```solidity
// Emitted when address record changes in dedicated resolver
// Note: node is always 0x0000... for alias support
event AddressChanged(
    bytes32 indexed node,
    uint256 coinType,
    bytes newAddress
);

event AddrChanged(
  bytes32 indexed node,
  address a
);

event TextChanged(
    bytes32 indexed node,
    string indexed indexedKey,
    string key,
    string value
);

event ContenthashChanged(
  bytes32 indexed node,
  bytes hash
);
```

> To cater for the alias feature enabled by the hierarchical registry model, node can be set as 0x, representing that the record is for any names that uses the resolver. 

Author: Aliasing is the by-product of hierarchical  registry system and the diagram below is to explain how aliasing looks like. I can add a sentense like this.

`As a by-product of hierarchical registry system, multiple registries can point to the same some registry.  For example both ".xyz" and ".eth" can point to "example" as their subnames, so that both "example.eth" and "example.xyz" point to the same registry and sharing the same resolver data. The following diagram shows the relationship of each registry and its dedicated resolver linked by multiple parent node to make the aliasing effect.`

    foo -.->|ResolverUpdate| resolver_foo["🔧 Dedicated Resolver<br/>Contract: 0x333...444<br/>node: 0x000...000 (alias mode)<br/>📍 address: 0xuser...2<br/>📝 text: purpose='foo specific'"]

    %% Styling
    classDef registry fill:#e1f5fe,stroke:#01579b,stroke-width:2px,color:#000,text-wrap:nowrap;
    classDef resolver fill:#fff8e1,stroke:#e65100,stroke-width:2px,color:#000,text-wrap:nowrap;
    classDef example fill:#f3e5f5,stroke:#4a148c,stroke-width:2px,color:#000,text-wrap:nowrap;
    class eth,xyz,foo registry;
    class example example;
    class resolver_example,resolver_foo resolver;
```

ENS v2 contract does not restrict from subregistry to point to one of its ancestor in the hierarchy causing circular structure. When indexing, it should stop indexing if a circular dependency is detected.

#### Indexer Implementation Notes

To prevent circular dependencies and redundant indexing:

1. **Track Visited Registries**: Maintain a set of visited registry addresses when traversing the hierarchy to detect cycles

> 2. **Skip Redundant Paths**: When a registry is reached through multiple parent registries (alias feature), index it only once

Reviewer: ```suggestion
2. **Skip Redundant Paths**: When a registry is reached through multiple parent registries via aliasing, index it only once
```

> 3. **Canonical Path Selection**: For aliased names, indexers should determine a canonical path (e.g., shortest path or first discovered path) to avoid duplicate entries

Reviewer: I think we should be more opinionated about this. Shortest path makes a lot of sense but I could see a world where that's not the desired effect from the user. Leaving this portion of the spec undefined will result in ambiguous indexing behavior.

Author: We can pick the shortest path unless others has different opinions

4. **Cycle Detection**: If a registry's subregistry points to any of its ancestors, stop traversal at that point to prevent infinite loops

Example indexing logic:
```
function indexRegistry(registry, visitedSet, path) {
    if (visitedSet.has(registry)) {
        // Circular dependency detected, stop indexing
        return;
    }
    visitedSet.add(registry);
    
    // Index current registry events
    // Then recursively index subregistries
    for (subregistry of registry.subregistries) {
        indexRegistry(subregistry, visitedSet, path + label);
    }
}
```

### Event Usage Patterns

#### ENS v2 Hierarchical Name Construction

To construct full domain names in the hierarchical model:

1. Start from the root registry and traverse down through subregistries
2. Use `SubregistryUpdate` events to find child registries for each label
3. Query `NewSubname` events at each registry level to get label information
4. Build the complete name by concatenating labels as you traverse down the hierarchy
5. Resolve final records using `ResolverUpdate` and resolver-specific events

> 6. Resolvers use `node = 0x0000...` to set record 

Author: Actually it's optional. I can rephrase to something like this.

`If resolver's node is set to 0x000, then indexer reconstructs the full node/namehash by traversing each label set on the registry hierarchy.`

#### Cross-Chain Name Resolution

For names pointing to L2 chains:

> 1. Name must first be ejected from L2 to L1 via `TransferSingle`, `TransferBatch`, `Transfer` event

Reviewer: What if the name already exists on L1? i.e. foo.eth already exists on L1 and migrates in place. Do we still get the requisite events? Or does migration explicitly mean that the name moves to L2 and then has to be ejected?

Author: How would you suggest it to be reworded to?

Author: 1. reworded to "bridged" 2. I don't understand what your actual question is here

#### Event Emission

In the simplified model:
- `NewSubname` events are emitted when creating subnodes
- `Transfer` events track ownership changes
- Resolver events use the full namehash as the node parameter
- No `SubregistryUpdate` or `ResolverUpdate` events are needed since registry and resolver are combined

#### Example Implementation

```solidity
contract SimplifiedL2Registry is ERC721, AddrResolver {
    mapping(bytes32 => bytes) public names;
    
    function createSubnode(
        bytes32 parentNode,
        string memory label
    ) external returns (bytes32 node) {
        bytes32 labelhash = keccak256(bytes(label));
        uint256 tokenId = uint256(labelhash);
        
        // Registry functionality: mint the NFT
        _mint(msg.sender, tokenId);
        
        // Calculate and store the full node
        node = keccak256(abi.encodePacked(parentNode, labelhash));
        names[node] = abi.encodePacked(names[parentNode], uint8(bytes(label).length), label);
        
        // Emit registry event
        emit NewSubname(labelhash, label);
        emit Transfer(address(0), msg.sender, tokenId);
    }
    

>     function setAddr(bytes32 node, address addr) external {

Reviewer: Example contract should use coinType param since this is required for ENSIP-19

---
description: L2 Registry and Resolver Events for Cross-Chain Name Services
contributors:
  - matoken.eth
ensip:
  status: draft
  created: 2025-08-04
---

# ENSIP-X: L2 Registry and Resolver Events for Cross-Chain Name Services

## Abstract

This ENSIP specifies standardized registry and resolver events for L2 and cross-chain name services that build upon ENSIP-16's metadata discovery mechanism. It defines events that support ENS v2's hierarchical registry model on NameChain and provides a framework for independent L2 subname issuing services through enhanced event-based indexing.

## Motivation

As ENS expands to L2 chains and adopts a hierarchical registry model, there is a need for standardized events that allow indexers to track name ownership, registry relationships, and resolver updates across chains. This ENSIP works in conjunction with ENSIP-16, which provides the discovery mechanism via the MetadataChanged event, to create a complete cross-chain name service infrastructure.

## Specification

### Events

#### Registry Events

```solidity
// Emitted when a new subname is registered

> event NewSubname(bytes32 indexed labelhash, string label);

Author: It has been discussed at https://github.com/ensdomains/namechain/pull/143

// Standard ERC1155 transfer event for name ownership changes
event TransferSingle(
    address indexed operator,
    address indexed from,
    address indexed to,
    uint256 id,
    uint256 value
);

// Standard ERC1155 transfer event for multiple name ownership changes
event TransferBatch(
  address indexed operator,
  address indexed from,
  address indexed to,
  uint256[] ids,
  uint256[] values

> )

Reviewer 2: ```suggestion
);
```

// Standard ERC721 transfer event for name ownership changes
event Transfer(
  address indexed from,
  address indexed to,

>   uint256 tokenId,
> )

Reviewer 2: ```suggestion
  uint256 tokenId
);
```

> // Emitted when subregistry is updated

Author: Added into the Terminology section

event SubregistryUpdate(

>     uint256 indexed id,

Author: Explained at the terminology section

    %% Styling
    classDef registry fill:#e1f5fe,stroke:#01579b,stroke-width:2px,color:#000,text-wrap:nowrap;
    classDef resolver fill:#fff8e1,stroke:#e65100,stroke-width:2px,color:#000,text-wrap:nowrap;
    classDef example fill:#f3e5f5,stroke:#4a148c,stroke-width:2px,color:#000,text-wrap:nowrap;
    class eth,xyz,foo registry;
    class example example;
    class resolver_example,resolver_foo resolver;
```

The ENS v2 contract does not restrict a subregistry from pointing to one of its ancestors in the hierarchy. This causes a circular structure. Indexers should stop indexing if a circular dependency is detected.

#### Indexer Implementation Notes

To prevent circular dependencies and redundant indexing:

1. **Track Visited Registries**: Maintain a set of visited registry addresses when traversing the hierarchy to detect cycles
2. **Skip Redundant Paths**: When a registry is reached through multiple parent registries via aliasing, index it only once
3. **Canonical Path Selection**: For aliased names, indexers should determine a canonical path with the shortest path to avoid duplicate entries

> 4. **Cycle Detection**: If a registry's subregistry points to any of its ancestors, stop traversal at that point to prevent infinite loops

Reviewer 3: Sure. The concern I have with the word "stop" is how it might be interpreted in the context of indexers as "crash".

Here's a quick draft of some alternative language but this also needs more work: "If a registry's subregistry points to any of its ancestors, the registry should be considered as being outside of the active ENS name tree".

This likely needs more work as I need to meditate on everything here more, but here's the key idea I'm trying to hint at: I believe indexers will need to separate the following ideas:
1. Indexing all events for all registries (specifically relative to the associated registry).
2. Indexing the "active ENS name tree".

In other words, as registries are moved from one parent to another, they may:
* Enter the "active ENS name tree"
* Leave the "active ENS name tree"
* Remain inside the "active ENS name tree"
* Remain outside the "active ENS name tree"

Here I'm defining the "active ENS name tree" as data records that are reachable through dynamic ENS resolution.

A related idea already exists within ENSv1 with resolver records. Ex: Nothing stops someone from creating a resolver that says the ETH deposit address for node X is Y, without any ownership or control over node X. But until the owner of node X makes this resolver the active resolver for node X, the resolver record stating that the ETH deposit address for node X is Y is outside the "active ENS name tree". This resolver record is "detached".

At any moment, this resolver record can become "attached" or "detached" as the resolvers assigned in the Registry are updated.

I'll need to switch my focus back to ENSv2 for several days to be able to articulate these ideas better. Currently I'm really focused on some refactoring we're doing in ENSNode.

Example indexing logic:

> ```

Reviewer 2: ```suggestion
```js
```

function indexRegistry(registry, visitedSet, path) {
    if (visitedSet.has(registry)) {
        // Circular dependency detected, stop indexing
        return;
    }
    visitedSet.add(registry);
    
    // Index current registry events
    // Then recursively index subregistries
    for (subregistry of registry.subregistries) {
        indexRegistry(subregistry, visitedSet, path + label);
    }
}
```

### Event Usage Patterns

#### ENS v2 Hierarchical Name Construction

To construct full domain names in the hierarchical model:

1. Start from the root registry and traverse down through subregistries
2. Use `SubregistryUpdate` events to find child registries for each label

> 3. Query `NewSubname` events at each registry level to get label information

Reviewer 3: Thanks. Rewording "Query" to "Use" helps.

Separately I'll work on a fresh top-to-bottom review of all the updates here 👍

4. Build the complete name by concatenating labels as you traverse down the hierarchy
5. Resolve final records using `ResolverUpdate` and resolver-specific events
6. If resolver's node is set to 0x000, then indexer reconstructs the full node/namehash by traversing each label set on the registry hierarchy.

#### Cross-Chain Name Resolution

For names pointing to L2 chains:

1. Name must first be ejected from L2 to L1 via `TransferSingle`, `TransferBatch`, `Transfer` event
2. L1 resolver emits `MetadataChanged` event (defined in ENSIP-16) with L2 chain information

> 3. Clients query the specified `graphqlUrl` for L2 name data

Reviewer 3: Thanks. Could you please clarify the framing of the narrative in ENSIP-X? As I understand from your recent message on Slack, ENSIP-X is targeted at indexers. However on this line it references "Clients". Is it correct to understand that this is exclusively a reference to "Indexers"? If so suggest to rename "Clients" to "Indexers".

Goal: "Indexers" are "Clients" too, but a specific subtype of "Client". As I understand the idea here is exclusive to "Indexers"

#### Overview

For simpler L2 implementations that don't require the full hierarchical registry architecture, contracts can combine registry and resolver functionality into a single contract. This approach is suitable for:
- Independent L2 subname services  
- Flat namespace implementations
- Lightweight name service deployments

#### Implementation Pattern

The simplified model:
1. Combines registry ownership tracking with resolver record storage
2. Emits both registry events (NewSubname, Transfer) and resolver events (AddressChanged, TextChanged, etc.)
3. Uses namehash for node identification instead of hierarchical traversal
4. Maintains compatibility with the standard events defined in this ENSIP

#### Event Emission

In the simplified model:
- `NewSubname` events are emitted when creating subnodes
- `Transfer` events track ownership changes
- Resolver events use the full namehash as the node parameter
- No `SubregistryUpdate` or `ResolverUpdate` events are needed since registry and resolver are combined

#### Example Implementation

```solidity
contract SimplifiedL2Registry is ERC721, AddrResolver {
    mapping(bytes32 => bytes) public names;
    
    function createSubnode(

>         bytes32 parentNode,

Author: ` will the subnames and resolver records previously added beneath project-a.eth continue to work now that they are beneath project-b.eth instead?` yes

        string memory label
    ) external returns (bytes32 node) {
        bytes32 labelhash = keccak256(bytes(label));
        uint256 tokenId = uint256(labelhash);
        
        // Registry functionality: mint the NFT
        _mint(msg.sender, tokenId);
        
        // Calculate and store the full node
        node = keccak256(abi.encodePacked(parentNode, labelhash));
        names[node] = abi.encodePacked(names[parentNode], uint8(bytes(label).length), label);
        
        // Emit registry event
        emit NewSubname(labelhash, label);
        emit Transfer(address(0), msg.sender, tokenId);
    }
    
    function setAddr(bytes32 node, uint256 coinType, address addr) external {
        require(isAuthorised(node), "Not authorized");
        
        // Resolver functionality: store the address
        addresses[node][coinType] = addr;
        
        // Emit resolver event with full namehash

>         emit AddrChanged(node, coinType, addr);

Reviewer 3: logic here should follow https://github.com/ensdomains/ens-contracts/blob/2429476746e7d577d3bf5c77c9979474fa4d47cd/contracts/resolvers/profiles/AddrResolver.sol#L59-L61

    }
}
```

#### Usage Pattern

1. Contract creates a subname and emits `NewSubname(labelhash, label)`
2. Same transaction can emit resolver events like `AddressChanged(namehash, coinType, address)`
3. Ownership transfers emit standard ERC721/ERC1155 Transfer events

> 4. All resolver updates use the full namehash for the node parameter

Reviewer 3: ```suggestion
4. All resolver updates use the namehash of the full name for the node parameter
```

### GraphQL Schema

> The following schema should be implemented by L2 indexers to provide a standard interface for querying name data:

Reviewer 3: Please see related feedback:

1. Strong recommendation against any use of GraphQL in ENSIPs
2. It's not correct to conflate the idea of indexer with offering a HTTP API that returns data / metadata about ENS names. These are independent ideas.
3. Why is the word "L2" necessary to use here? Is it therefore guaranteed these ideas never apply to anything on L1?
